nomad/job_endpoint.go

```diff
@@ -1298,19 +1298,65 @@ func (j *Job) GetJobVersions(args *structs.JobVersionsRequest,
 			// Setup the output
 			reply.Versions = out
 			if len(out) != 0 {
+
+				var compareVersionNumber uint64
+				var compareVersion *structs.Job
+				var compareStatic bool
+
+				if args.Diffs {
+					if args.DiffTagName != "" {
+						compareStatic = true
+						tagFound := false
+						for _, version := range out {
+							if version.TaggedVersion != nil && version.TaggedVersion.Name == args.DiffTagName {
+								tagFound = true
+								compareVersionNumber = version.Version
+								break
+							}
+						}
+						if !tagFound {
+							return fmt.Errorf("tag %q not found", args.DiffTagName)
+						}
+					} else if args.DiffVersion != nil {
+						compareStatic = true
+						compareVersionNumber = *args.DiffVersion
+					}
+				}
+
 				// Note: a previous assumption here was that the 0th job was the latest, and that we don't modify "old" versions.
 				// Adding version tags breaks this assumption (you can tag an old version, which should unblock /versions queries) so we now look for the highest ModifyIndex.
 				var maxModifyIndex uint64
 				for _, job := range out {
 					if job.ModifyIndex > maxModifyIndex {
 						maxModifyIndex = job.ModifyIndex
 					}
+					if compareStatic && job.Version == compareVersionNumber {
+						compareVersion = job
+					}
 				}
 				reply.Index = maxModifyIndex
+
+				if compareStatic && compareVersion == nil {
+					if args.DiffTagName != "" {
+						return fmt.Errorf("tag %q not found", args.DiffTagName)
+					}
+					return fmt.Errorf("version %d not found", *args.DiffVersion)
+				}
+
 				// Compute the diffs
 				if args.Diffs {
-					for i := 0; i < len(out)-1; i++ {
-						old, new := out[i+1], out[i]
+					for i := 0; i < len(out); i++ {
+						// For each of your job's versions, counting from highest version to lowest
+						// it should either be compared to the NEXT version in out[], or to the pre-specified version from compareStatic.
+						if !compareStatic {
+							if i+1 < len(out) && out[i+1] != nil {
+								compareVersion = out[i+1]
+							} else {
+								break
+							}
+						}
+
+						old, new := compareVersion, out[i]
```

Comment:
Unconditionally assigning from the slice lets you avoid checking the length twice and doing the nil check (although I'm not sure that actually does anything anyways). We can tighten this up a bit like so:

```suggestion
					for i := 0; i < len(out)-1; i++ {
						// For each of your job's versions, counting from highest version to lowest
						// it should either be compared to the NEXT version in out[], or to the pre-specified version from compareStatic.

						old, new := out[i+1], out[i]
						if compareStatic {
							old = compareVersion
						}
```